![image](https://user-images.githubusercontent.com/24732704/55021982-f822ec00-4ff9-11e9-802a-649cfdb4892c.png)

## UMC Utrecht internship code repository - Annotate rsIDs for SNPs

### Background
GWAS summary statistics usually include rsIDs for SNPs in the 'SNP' column, however, you might need to annotate them in some cases. This script uses a SNP annotation table obtained from UCSC Table Browser [1] (see _UCSC Data_) to annotate rsIDs to SNPs for summary statistics. It includes a specific step to remove allele details from the SNP column, which -depending on data format- might not be necessary. Additionally, only items with a length of 1bp are included, thus excluding longer insertions.

### Scripts
- **snp_annotate_rs_ids.sh** - Script that annotates rsIDs for SNPs.

### UCSC Data
The following parameters were used to obtain the SNP annotation table from [UCSC Table Browser](https://genome.ucsc.edu/cgi-bin/hgTables):

| Parameter          | Value(s)                                      |
|:-------------------|:----------------------------------------------|
| Clade              | Mammal                                        |
| Genome             | Human                                         |
| Assembly           | Feb. 2009 (GRCh37/hg19)                       |
| Group              | Variation                                     |
| Track              | All SNPs (151)                                |
| Region             | Genome                                        |
| Output format      | Select fields from primary and related tables |
| File type returned | Plain text                                    |
| Selected fields     | chrom, chromStart, chromEnd, name             |

Note: This file easily exceeds 12 GB!

### Data
- UCSC SNP annotation table
- GWAS summary statistics

### Usage

1. `chmod 755 snp_annotate_rs_ids.sh`
1. Set file paths in the script
1. `./snp_annotate_rs_ids.sh`
1. Output can be appended to your original summary statistics as you please

### Requirements
- Unix environment (I used Ubuntu 18.04), including AWK:
`sudo apt-get install gawk`

### References
[1] Karolchik et al. (2004), "The UCSC Table Browser data retrieval tool", _Nucleic Acids Research_ 32, pD493â€“D496, doi: https://doi.org/10.1093/nar/gkh103

### Glossary
* **SNP** - Single base-pair change that occurs on a specific genome position
* **rsID** - Reference SNP ID, identfier for SNPs that map to identical genomic position, assigned by NCBI to reduce redundancy
